dask/bag/core.py

```diff
@@ -751,6 +758,90 @@ def write(data, filename):
         f.close()
 
 
+def from_s3(bucket_name, paths='*', aws_access_key=None, aws_secret_key=None,
+            connection=None, anon=False):
+    """ Create a dask.bag by loading files from s3.
+
+    >>> b = from_s3(bucket, 'myfile1.txt')
+
+    or a pattern with '*' or '?'. There is one partition per file.
+
+    >>> b = from_s3(bucket, '*.json')
+    >>> b.npartitions
+    9
+
+    or a list of files.
+
+    >>> b = from_s3(bucket, ['alice.csv', 'bob.csv'])
+    >>> b.npartitions
+    2
+
+    """
+    def _connect_s3(aws_access_key, aws_secret_key, connection, anon):
+        """Connect to s3"""
+        import boto
+        if anon is True:
+            connection = boto.connect_s3(anon=anon)
+        elif connection is None:
+            connection = boto.connect_s3(aws_access_key, aws_secret_key)
+        return connection
+
+    def _from_s3(bucket_name, matches, conn_args):
+        """
+        takes an s3 bucket object and a list of bucket strings
+        """
+        dsk = {}
+        for i, m in enumerate(matches):
+            # connect to s3
+            a, b, c, d = conn_args
+            get_connection = (_connect_s3, a, b, c, d)
+            # get the bucket with the connection
+            get_bucket = (lambda x: x.get_bucket(bucket_name), get_connection)
+            # get a key from the bucket
+            get_key = (lambda b, k: b.get_key(k), get_bucket, m)
+            # read the key
+            read_key = lambda k: [getattr(k, 'read')()]
+            dsk[('load', i)] = (read_key, get_key)
+        return Bag(dsk, 'load', len(matches))
+
+    conn_args = (aws_access_key, aws_secret_key, connection, anon)
+
+    if bucket_name.startswith('s3://'):
+        # FIXME
+        # DIRTY HACK: so that urlparse won't think '?' in the URI path is for a
+        # query. We replace '?' with 'QUESTIONMARK' so it stays in the path
+        # then after urlpars gives us the path replace 'QUESTIONMARK' with '?'.
+        if '?' in bucket_name:
+            bucket_name = bucket_name.replace('?', 'QUESTIONMARK')
+        o = urlparse(bucket_name)
+        # if path is specified
+        if (paths == '*') and (o.path != '' and o.path != '/'):
+            paths = o.path[1:].replace('QUESTIONMARK', '?')
+        bucket_name = o.hostname
+
+    # get the bucket
+    conn = _connect_s3(*conn_args)
+    bucket = conn.get_bucket(bucket_name)
+
+    # sing bucket key, or pattern to expand
+    if isinstance(paths, str):
+        if ('*' not in paths) and ('?' not in paths):
+            return _from_s3(bucket_name, [paths], conn_args)
+
+        # handle globs
+        keys = bucket.list()
+
+        matches = []
+        for k in keys:
+            k_name = k.name.encode('utf-8')
+            if fnmatchcase(k_name, paths):
+                matches.append(k_name)
+        return _from_s3(bucket_name, matches, conn_args)
```

Comment:
Is there a way to isolate the hacky code into a separate testable function?  This lets us test it more heavily in isolation and isolates the hackiness away from general S3 loading, which is ideally somewhat striaghtforward.